src/System.Management.Automation/engine/InternalCommands.cs

```diff
@@ -203,69 +251,241 @@ public object[] ArgumentList
+                    break;
+            }
+        }
+
+        /// <summary>
+        /// Execute the end scriptblock when the pipeline is complete.
+        /// </summary>
+        /// <exception cref="ParseException">Could not parse script.</exception>
+        /// <exception cref="RuntimeException">See Pipeline.Invoke.</exception>
+        /// <exception cref="ParameterBindingException">See Pipeline.Invoke.</exception>
+        protected override void EndProcessing()
+        {
+            switch (ParameterSetName)
+            {
+                case "ScriptBlockSet":
+                    EndBlockParameterSet();
+                    break;
+
+                case ForEachObjectCommand.ParallelParameterSet:
+                    EndParallelParameterSet();
+                    break;
+            }
+        }
+
+        /// <summary>
+        /// Handle pipeline stop signal.
+        /// </summary>
+        protected override void StopProcessing()
+        {
+            switch (ParameterSetName)
+            {
+                case ForEachObjectCommand.ParallelParameterSet:
+                    StopParallelProcessing();
+                    break;
+            }
+        }
+
+        #endregion
+
+        #region IDisposable
+
+        /// <summary>
+        /// Dispose cmdlet instance.
+        /// </summary>
+        public void Dispose()
+        {
+            // Ensure all parallel task objects are disposed
+            if (_taskTimer != null)
+            {
+                _taskTimer.Dispose();
+            }
+            if (_taskDataStreamWriter != null)
+            {
+                _taskDataStreamWriter.Dispose();
+            }
+            if (_taskPool != null)
+            {
+                _taskPool.Dispose();
+            }
+        }
+
+        #endregion
+        
+        #region Private Methods
+
+        #region PSTasks
+
+        private PSTaskPool _taskPool;
+        private PSTaskDataStreamWriter _taskDataStreamWriter;
+        private Dictionary<string, object> _usingValuesMap;
+        private System.Threading.Timer _taskTimer;
+        private PSTaskJob _taskJob;
+
+        private void InitParallelParameterSet()
+        {
+            bool allowUsingExpression = this.Context.SessionState.LanguageMode != PSLanguageMode.NoLanguage;
+            _usingValuesMap = ScriptBlockToPowerShellConverter.GetUsingValuesAsDictionary(Parallel, allowUsingExpression, this.Context, null);
+            
+            // Validate using values map
+            foreach (var item in _usingValuesMap.Values)
+            {
+                if (item is ScriptBlock)
                 {
-                    whatIf = (SwitchParameter)argument;
+                    this.ThrowTerminatingError(
+                        new ErrorRecord(
+                            new PSArgumentException(InternalCommandStrings.ParallelUsingVariableCannotBeScriptBlock),
+                            "ParallelUsingVariableCannotBeScriptBlock",
```

Comment:
Shoudl this string be in a resx?

Reply:
No, fully qualified error id strings are not localized.